<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.6.39">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Joseph Rickert">
<meta name="dcterms.date" content="2024-11-25">
<meta name="description" content="One hundred eighty-one new packages made CRAN’s final cut in October.">

<title>October 2024: Top 40 New CRAN Packages – R Works</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
</style>


<script src="../../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../../site_libs/quarto-search/fuse.min.js"></script>
<script src="../../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../../">
<script src="../../site_libs/quarto-html/quarto.js"></script>
<script src="../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../../site_libs/quarto-html/anchor.min.js"></script>
<link href="../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting-05fe91a66cf75bbbb8c9664867fe5124.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../../site_libs/bootstrap/bootstrap-e4bce21b2273d57a326e73fb72c29580.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">
<script src="../../site_libs/quarto-contrib/bluesky-comments-section-1.0.0/bluesky-comments.js"></script>
<script src="../../site_libs/quarto-contrib/iconify-2.1.0/iconify-icon.min.js"></script>
<script src="../../site_libs/quarto-contrib/glightbox/glightbox.min.js"></script>
<link href="../../site_libs/quarto-contrib/glightbox/glightbox.min.css" rel="stylesheet">
<link href="../../site_libs/quarto-contrib/glightbox/lightbox.css" rel="stylesheet">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>
<style>

      .quarto-title-block .quarto-title-banner {
        background: #FFF;
      }
</style>
<script defer="" data-domain="rworks.dev" src="https://plausible.io/js/script.js"></script>


<meta property="og:title" content="October 2024: Top 40 New CRAN Packages – R Works">
<meta property="og:description" content="One hundred eighty-one new packages made CRAN’s final cut in October.">
<meta property="og:image" content="https://rworks.dev/posts/october-2024-top-40-new-cran-packages/clustTMB.png">
<meta property="og:site_name" content="R Works">
<meta property="og:image:height" content="384">
<meta property="og:image:width" content="576">
<meta name="twitter:title" content="October 2024: Top 40 New CRAN Packages – R Works">
<meta name="twitter:description" content="One hundred eighty-one new packages made CRAN’s final cut in October.">
<meta name="twitter:image" content="https://rworks.dev/posts/october-2024-top-40-new-cran-packages/clustTMB.png">
<meta name="twitter:image-height" content="384">
<meta name="twitter:image-width" content="576">
<meta name="twitter:card" content="summary_large_image">
</head>

<body class="nav-fixed fullcontent">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top quarto-banner">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="../../index.html">
    <span class="navbar-title">R Works</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" role="menu" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="../../about.html"> 
<span class="menu-text">About</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../how-to-contribute.html"> 
<span class="menu-text">How to Contribute</span></a>
  </li>  
</ul>
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item">
    <a class="nav-link" href="https://bsky.app/profile/rworks.bsky.social"> 
<span class="menu-text"><iconify-icon role="img" inline="" icon="simple-icons:bluesky" aria-label="Icon bluesky from simple-icons Iconify.design set." title="Icon bluesky from simple-icons Iconify.design set."></iconify-icon></span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="https://rworks.dev/index.xml"> <i class="bi bi-rss" role="img" aria-label="R Works Blog RSS">
</i> 
<span class="menu-text"></span></a>
  </li>  
</ul>
          </div> <!-- /navcollapse -->
            <div class="quarto-navbar-tools">
</div>
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<header id="title-block-header" class="quarto-title-block default page-columns page-full">
  <div class="quarto-title-banner page-columns page-full">
    <div class="quarto-title column-body">
      <h1 class="title">October 2024: Top 40 New CRAN Packages</h1>
                  <div>
        <div class="description">
          One hundred eighty-one new packages made CRAN’s final cut in October.
        </div>
      </div>
                          <div class="quarto-categories">
                <div class="quarto-category">Top 40</div>
              </div>
                  </div>
  </div>
    
  
  <div class="quarto-title-meta">

      <div>
      <div class="quarto-title-meta-heading">Author</div>
      <div class="quarto-title-meta-contents">
               <p>Joseph Rickert </p>
            </div>
    </div>
      
      <div>
      <div class="quarto-title-meta-heading">Published</div>
      <div class="quarto-title-meta-contents">
        <p class="date">November 25, 2024</p>
      </div>
    </div>
    
      
    </div>
    
  
  </header><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    
<!-- main -->
<main class="content quarto-banner-title-block" id="quarto-document-content">





<p>One hundred eighty-one new packages made CRAN’s final cut in October. Here are my Top 40 picks in thirteen categories: AI, Climate Analysis, Computational Methods, Data, Epidemiology, Genomics, Machine Learning, Medicine, Quality Management, Statistics, Time Series, Utilities, and Visualization.</p>
<div class="columns">
<div class="column" style="width:45%;">
<section id="ai" class="level3">
<h3 class="anchored" data-anchor-id="ai">AI</h3>
<p><a href="https://cran.r-project.org/package=hollr">hollr</a> v1.0.0: Enables chat completion and text annotation with local and <a href="https://openai.com/">OpenAI</a> language models and supports batch processing, multiple annotators, and consistent output formats. See <a href="https://cran.r-project.org/web/packages/hollr/readme/README.html">README</a>.</p>
<p><a href="https://cran.r-project.org/package=spBPS">spBPS</a> v0.0-4: Provides functions for Bayesian Predictive Stacking within the Bayesian transfer learning framework for geospatial artificial systems, as described in <a href="https://arxiv.org/abs/2410.09504">Rewaiccw &amp; Banerjee (2024)</a>. Core functions leverage <code>C++</code>, making the framework well-suited for large-scale spatial data analysis in parallel and distributed computing environments.</p>
<p><a href="spBPS.png" class="lightbox" data-gallery="quarto-lightbox-gallery-1"><img src="spBPS.png" class="img-fluid" alt="Heatmaps of Spacial Processes"></a></p>
<p><a href="https://cran.r-project.org/package=tidyllm">tidyllm</a> v0.1.0: Implements a tidy interface for integrating large language model (LLM) APIs such as <a href="https://www.anthropic.com/">Claude</a>, <a href="https://openai.com/">ChatGPT</a>, <a href="https://groq.com/">Groq</a>, and local models via <a href="https://ollama.com/">Ollama</a> into R workflows and supports text and media-based interactions, interactive message history, stateful rate limit handling, and a tidy, pipeline-oriented interface for streamlined integration into data workflows. See the <a href="https://cran.r-project.org/web/packages/tidyllm/vignettes/getting-started.html">vignette</a>.</p>
</section>
<section id="climate-analysis" class="level3">
<h3 class="anchored" data-anchor-id="climate-analysis">Climate Analysis</h3>
<p><a href="https://cran.r-project.org/package=carbonr">carbonr</a> v0.2.1: Provides a tool for calculating carbon-equivalent emissions based on the UK Government’s <a href="https://www.gov.uk/government/publications/greenhouse-gas-reporting-conversion-factors-2023">Greenhouse Gas Conversion Factors</a> report; it facilitates transparent emissions calculations for various sectors, including travel, accommodation, and clinical activities. See the <a href="https://cran.r-project.org/web/packages/carbonr/vignettes/introduction-to-carbonr.html">vignette</a>.</p>
<p><a href="https://cran.r-project.org/package=fluxfinder">fluxfinder</a> v1.0.0: Parse static-chamber greenhouse gas measurement files generated by a variety of instruments; compute flux rates using multi-observation metadata; and generate diagnostic metrics and plots. Designed to be easy to integrate into reproducible scientific workflows. There is an <a href="https://cran.r-project.org/web/packages/fluxfinder/vignettes/intro-to-fluxfinder.html">Introduction</a> and a <a href="https://cran.r-project.org/web/packages/fluxfinder/vignettes/intro-to-fluxfinder.html">vignette</a> on integrating with the <code>gasfluxes</code> package.</p>
<p><a href="fluxfinder.png" class="lightbox" data-gallery="quarto-lightbox-gallery-2"><img src="fluxfinder.png" class="img-fluid" alt="Time series colored by metadata"></a></p>
</section>
<section id="computational-methods" class="level3">
<h3 class="anchored" data-anchor-id="computational-methods">Computational Methods</h3>
<p><a href="https://cran.r-project.org/package=gmresls">gmresls</a> v0.2.2: Implements a method to solve a least squares system Ax~=b (dim(A)=(m,n) with m &gt;= n) with a precondition matrix B: BAx=Bb (dim(B)=(n,m)) based on the General Minimal Residual Algorithm of <a href="https://epubs.siam.org/doi/10.1137/0907058">Saad &amp; Schultz (1986)</a>. See <a href="https://cran.r-project.org/web/packages/gmresls/readme/README.html">README</a>.</p>
<p><a href="https://cran.r-project.org/package=mappeR">mappeR</a> v1.3.0: Implements an algorithm that generalizes the concept of a <a href="https://en.wikipedia.org/wiki/Reeb_graph">Reeb Graph</a> as described in <a href="https://research.math.osu.edu/tgda/mapperPBG.pdf">Mémoli and Carlsson (2007)</a> for a Topological Data Analysis of high dimensional data. Look <a href="https://github.com/Uiowa-Applied-Topology/mappeR">here</a> for examples.</p>
<p><a href="mapper.png" class="lightbox" data-gallery="quarto-lightbox-gallery-3"><img src="mapper.png" class="img-fluid" alt="Example of mappeR Graph"></a></p>
<p><a href="https://CRAN.R-project.org/package=TensorTools">TensorTools</a> v1.0.0: Provides a set of tools for basic tensor operators based on the Discrete Fourier Transform, including the eigenvalue, QR decomposition, LU decompositions of a tensor, and the calculation of the inverse of a tensor and the transpose of a symmetric tensor. See <a href="https://www.sciencedirect.com/science/article/pii/S0024379515004358?via%3Dihub">Kernfeld et al.&nbsp;(2015)</a> for the details and the <a href="https://cran.r-project.org/web/packages/TensorTools/vignettes/TensorTools.html">vignette</a> for examples. Note, in this context, a tensor is a multidimensional array.</p>
</section>
<section id="data" class="level3">
<h3 class="anchored" data-anchor-id="data">Data</h3>
<p><a href="https://cran.r-project.org/package=openFDA">openFDA</a> v0.1.0: Facilitates access to U.S. Food and Drug Administration’s <a href="https://open.fda.gov/">openFDA</a> data on drugs, devices, foodstuffs, tobacco, and more. See <a href="https://academic.oup.com/jamia/article/23/3/596/2908999?login=false">Kass-Hout et al.&nbsp;(2016)</a> for background and the <a href="https://cran.r-project.org/web/packages/openFDA/vignettes/openFDA.html">vignette</a> to get started.</p>
<p><a href="https://cran.r-project.org/package=ozbabynames">ozbabynames</a> v0.1.0: Provides data on the most popular baby names by sex and year for each state in Australia as provided by the state and territory governments. The quality and quantity of the data varies with the state. Look <a href="https://pkg.robjhyndman.com/ozbabynames/">here</a> for information.</p>
<p><a href="ozbaby.gif" class="lightbox" data-gallery="quarto-lightbox-gallery-4"><img src="ozbaby.gif" class="img-fluid" alt="Gif of bubs dubbed"></a></p>
</section>
<section id="epidemiology" class="level3">
<h3 class="anchored" data-anchor-id="epidemiology">Epidemiology</h3>
<p><a href="https://cran.r-project.org/package=epichains">epichains</a> v0.1.1: Provides methods to simulate and analyze the size and length of branching processes with an arbitrary offspring distribution. These can be used, for example, to analyze the distribution of chain sizes or length of infectious disease outbreaks, as discussed in <a href="https://academic.oup.com/biostatistics/article-abstract/4/2/279/245106?redirectedFrom=fulltext&amp;login=false">Farrington et al.&nbsp;(2003)</a>. There are six vignettes, including a <a href="https://cran.r-project.org/web/packages/epichains/vignettes/epichains.html">Getting Started</a> guide and <a href="https://cran.r-project.org/web/packages/epichains/vignettes/theoretical_background.html">Theoretical Background</a>.</p>
<p><a href="epichains.png" class="lightbox" data-gallery="quarto-lightbox-gallery-5"><img src="epichains.png" class="img-fluid" alt="Diagram depicting chain"></a></p>
<p><a href="https://cran.r-project.org/package=epizootic">epizootic</a> v1.0.0: Extends the pattern-oriented modeling framework of the <code>poems</code> package to provide functions for modeling disease transmission on a population scale in a spatiotemporally explicit manner and includes seasonal time steps, dispersal functions, objects that store disease states, and a population simulator that includes disease dynamics. See the <a href="https://cran.r-project.org/web/packages/epizootic/vignettes/mycoplasma.html">vignette</a> for an extended example.</p>
<p><a href="epizootic.png" class="lightbox" data-gallery="quarto-lightbox-gallery-6"><img src="epizootic.png" class="img-fluid" alt="Map of susceptible population"></a></p>
<p><a href="https://cran.r-project.org/package=serosv">serosv</a> v1.0.1: Implements tools to estimate infectious diseases parameters using serological data. Implemented models include SIR models, nonparametric models, semiparametric models, and hierarchical models, which are based on the book by <a href="https://link.springer.com/book/10.1007/978-1-4614-4072-7">Hens et al.&nbsp;(2013)</a>. There are eight vignettes including <a href="">Hierarchical Bayesian Models</a> and <a href="https://cran.r-project.org/web/packages/serosv/vignettes/model_quantitative_data.html">Modeling directly from antibody levels</a>.</p>
<p><a href="serosv.png" class="lightbox" data-gallery="quarto-lightbox-gallery-7"><img src="serosv.png" class="img-fluid" alt="Model visualization showing sero prevalence and fprce of infection"></a></p>
</section>
<section id="genomics" class="level3">
<h3 class="anchored" data-anchor-id="genomics">Genomics</h3>
<p><a href="https://cran.r-project.org/package=GSEMA">GSEMA</a> v0.99.3: Provides functions to perform various steps of gene set enrichment meta-analysis, including meta-analysis of effect sizes from different pathways in different studies. See the <a href="https://cran.r-project.org/web/packages/GSEMA/vignettes/GSEMA.html">vignette</a> for examples.</p>
<p><a href="GSEMA.png" class="lightbox" data-gallery="quarto-lightbox-gallery-8"><img src="GSEMA.png" class="img-fluid" alt="Heatmap of random effects model"></a></p>
<p><a href="https://cran.r-project.org/package=phylotypr">phylotypr</a> v0.1.0: Implements Naive Bayesian Classifier from the Ribosomal Database Project that traditionally has been used to classify 16S rRNA gene sequences to bacterial taxonomic outlines but which applies to any type of gene sequence. See <a href="https://journals.asm.org/doi/10.1128/aem.00062-07">Wang et al.&nbsp;(2007)</a> for background and the <a href="https://cran.r-project.org/web/packages/phylotypr/vignettes/phylotypr.html">vignette</a> for examples.</p>
</section>
<section id="machine-learning" class="level3">
<h3 class="anchored" data-anchor-id="machine-learning">Machine Learning</h3>
<p><a href="https://cran.r-project.org/package=textpress">textpress</a> v1.0.0: Provides a Natural Language Processing (NLP) toolkit focused on search-centric workflows with minimal dependencies. The package offers key features for web scraping, text processing, corpus search, and text embedding generation via the <a href="https://huggingface.co/docs/api-inference/index">HuggingFace API</a>. See <a href="https://cran.r-project.org/web/packages/textpress/readme/README.html">README</a> for examples.</p>
</section>
<section id="medicine" class="level3">
<h3 class="anchored" data-anchor-id="medicine">Medicine</h3>
<p><a href="https://cran.r-project.org/package=eyetools">eyetools</a> v0.7.2: Provides functions to help researchers analyze eye data enabling the automation of actions across a pipeline that includes transforming binocular data, gap repair, and event-based processing such as fixations, saccades, and entry and duration in areas of interest. Functions implement the fixation and saccade detection methods proposed by <a href="https://dl.acm.org/doi/10.1145/355017.355028">Salvucci and Goldberg (2000)</a> and visualize eye movement. See the <a href="https://cran.r-project.org/web/packages/eyetools/vignettes/eyetools.html">Introduction</a>.</p>
<p><a href="eyetools.png" class="lightbox" data-gallery="quarto-lightbox-gallery-9"><img src="eyetools.png" class="img-fluid" alt="Eye movement plots"></a></p>
<p><a href="https://cran.r-project.org/package=MedDataSets">MedDataSets</a> v0.1.0: Provides an extensive collection of datasets related to medicine, diseases, treatments, drugs, and public health, covering topics such as drug effectiveness, vaccine trials, survival rates, infectious disease outbreaks, and medical treatments. including AIDS, cancer, bacterial infections, and COVID-19, and information on pharmaceuticals and vaccines. These datasets are sourced from the R ecosystem and other R packages. See the <a href="https://cran.r-project.org/web/packages/MedDataSets/vignettes/intro_meddatasets.html">vignette</a>.</p>
<p><a href="https://cran.r-project.org/package=ODT">ODT</a> v1.0.0: Implements a tree-based method specifically designed for personalized medicine applications that uses genomic and mutational data to identify optimal drug recommendations tailored to individual patient profiles. See <a href="https://academic.oup.com/bib/article/24/4/bbad200/7186396?login=false">Gimeno et al.&nbsp;(2023)</a> for the details and the <a href="https://cran.r-project.org/web/packages/ODT/vignettes/ODT.html">vignette</a> for an example.</p>
<p><a href="ODT.png" class="lightbox" data-gallery="quarto-lightbox-gallery-10"><img src="ODT.png" class="img-fluid" alt="Decision Tree"></a></p>
<p><a href="https://cran.r-project.org/package=trtswitch">trtswitch</a> v0.1.1: Implements rank-preserving structural failure time model (RPSFTM), iterative parameter estimation (IPE), inverse probability of censoring weights (IPCW), and two-stage estimation (TSE) methods for treatment switching in randomized clinical trials. See <a href="https://pubmed.ncbi.nlm.nih.gov/25416688/">Latimer et al.&nbsp;(2017)</a> for background. There are five vignettes including descriptions of <a href="https://cran.r-project.org/web/packages/trtswitch/vignettes/rpsftm.html">RPSFTM</a> and <a href="https://cran.r-project.org/web/packages/trtswitch/vignettes/ipe.html">IPE</a>.</p>
<p><a href="https://cran.r-project.org/package=UnplanSimon">UnplanSimon</a> v0.1.0: Implements methods to manage under- and over-enrollment in Simon’s Two-Stage Design by providing adaptive threshold adjustments and sample size recalibration, and post-inference analysis tools to support clinical trial design and evaluation. See the <a href="https://cran.r-project.org/web/packages/UnplanSimon/vignettes/UnplanSimon.html">vignette</a>.</p>
</section>
<section id="quality-management" class="level3">
<h3 class="anchored" data-anchor-id="quality-management">Quality Management</h3>
<p><a href="https://cran.r-project.org/package=r6qualitytools">r6qualitytools</a> v1.0.1: Implements a suite of statistical tools for Quality Management, designed around the Define, Measure, Analyze, Improve, and Control (DMAIC) cycle used in Six Sigma methodology. It refactors the <code>qualitytools</code>package incorporating ‘R6’ object-oriented programming for increased flexibility and performance and replaces traditional graphics with modern, interactive visualizations. Look <a href="https://github.com/Fabianenc/r6qualitytool">here</a> for examples.</p>
<p><a href="r6quality.png" class="lightbox" data-gallery="quarto-lightbox-gallery-11"><img src="r6quality.png" class="img-fluid" alt="Process controol charts"></a></p>
</section>
<section id="statistics" class="level3">
<h3 class="anchored" data-anchor-id="statistics">Statistics</h3>
<p><a href="https://cran.r-project.org/package=clustTMB">clusTMB</a>: v0.1.0: Fits a spatio-temporal finite mixture model using the <code>TMB</code> package.. Covariate, spatial, and temporal random effects can be incorporated into the gating formula using multinomial logistic regression, the expert formula using a generalized linear mixed model framework, or both. See the vignettes <a href="https://cran.r-project.org/web/packages/clustTMB/vignettes/CovarianceStructure.html">Covariance Structure</a> and <a href="https://cran.r-project.org/web/packages/clustTMB/vignettes/SpatialExMeuseData.html">Meuse Example</a>.</p>
<p><a href="clustTMB.png" class="lightbox" data-gallery="quarto-lightbox-gallery-12"><img src="clustTMB.png" class="img-fluid" alt="Correlation plots of concentration levels"></a></p>
</section>
</div><div class="column" style="width:10%;">

</div><div class="column" style="width:45%;">
<section id="statistics-continued" class="level3">
<h3 class="anchored" data-anchor-id="statistics-continued">Statistics (Continued)</h3>
<p><a href="https://cran.r-project.org/package=extrememix">extrememix</a> v0.0.1: Fits extreme value mixture models, which are models for tails not requiring selection of a threshold, for continuous data. See <a href="https://journals.sagepub.com/doi/10.1191/1471082X04st075oa">Behrens et al.(2004)</a> and <a href="https://link.springer.com/article/10.1007/s11222-011-9270-z">Nascimento et al.&nbsp;(2011)</a> for the theory and the <a href="https://cran.r-project.org/web/packages/extrememix/vignettes/my-vignette.html">vignette</a> for examples.</p>
<p><a href="extrememix.png" class="lightbox" data-gallery="quarto-lightbox-gallery-13"><img src="extrememix.png" class="img-fluid" alt="Plot of Predictive Distribution"></a></p>
<p><a href="https://cran.r-project.org/package=MECfda">MECfda</a> v0.1.0: Implements functions to solve scalar-on-function linear models, including generalized linear mixed effect model and quantile linear regression model, and bias correction estimation methods due to measurement error. Details about the measurement error bias correction methods. See <a href="https://arxiv.org/abs/2305.12624">Luan et al.&nbsp;(2023)</a>, <a href="https://academic.oup.com/biostatistics/article/23/4/1218/6596004?login=false">Tekwe et al.&nbsp;(2022)</a>, <a href="https://www3.stat.sinica.edu.tw/statistica/J33N3/J33N321/J33N321.html">Zhang et al.&nbsp;(2023)</a>, and <a href="https://onlinelibrary.wiley.com/doi/10.1002/sim.8179">Tekwe et al.&nbsp;(2019)</a> for background and the <a href="https://cran.r-project.org/web/packages/MECfda/vignettes/MECfda.html">vignette</a> for an introduction.</p>
<p><a href="https://cran.r-project.org/package=multpois">multipois</a> v0.2.0: Implements a method to analyze polytomous responses with three or more unordered categories by transforming nominal response data into counts for each categorical alternative. These counts are then analyzed using (mixed) Poisson regression as per <a href="https://www.jstor.org/stable/2348134?origin=crossref">Baker (1994)</a>.</p>
<p><a href="multipois.png" class="lightbox" data-gallery="quarto-lightbox-gallery-14"><img src="multipois.png" class="img-fluid" alt="Mosaic plot of Y by, X1, X2"></a></p>
<p><a href="https://cran.r-project.org/package=spStack">spStack</a> v1.0.1: Fits Bayesian hierarchical spatial process models for point-referenced Gaussian, Poisson, binomial, and binary data using stacking of predictive densities by sampling from analytically available posterior distributions conditional upon some candidate values of the spatial process parameters. See <a href="https://arxiv.org/abs/2304.12414">Zhang et al.&nbsp;(2024)</a> and <a href="https://arxiv.org/abs/2406.04655">Pan et al.&nbsp;(2024)</a> for background and the <a href="https://cran.r-project.org/web/packages/spStack/vignettes/spStack.html">vignette</a> for examples.</p>
<p><a href="spStack.png" class="lightbox" data-gallery="quarto-lightbox-gallery-15"><img src="spStack.png" class="img-fluid" alt="Spatial Surface Plots"></a></p>
<p><a href="https://cran.r-project.org/package=svyROC">svyROC</a> v1.0.0: Provides functions to estimate the receiver operating characteristic (ROC) curve, area under the curve (AUC), and optimal cut-off points for individual classification, taking into account complex sampling designs when working with complex survey data. See <a href="https://onlinelibrary.wiley.com/doi/10.1002/sta4.635">Iparragirre et al.&nbsp;(2024)</a>, <a href="https://www.idescat.cat/sort/sort461/46.1.6.Iparragirre-etal.pdf">Iparragirre et al.&nbsp;(2022)</a> and <a href="https://link.springer.com/chapter/10.1007/978-3-031-65723-8_7">Iparragirre &amp; Barrio (2024)</a> for the theory and <a href="https://cran.r-project.org/web/packages/svyROC/readme/README.html">README</a> for an example.</p>
<p><a href="svyROC.png" class="lightbox" data-gallery="quarto-lightbox-gallery-16"><img src="svyROC.png" class="img-fluid" alt="ROC Curve"></a></p>
<p><a href="https://cran.r-project.org/package=VDPO">VDPO</a> v0.1.0: Provides a comprehensive set of tools for analyzing and manipulating functional data with non-uniform lengths and addresses two common scenarios in functional data analysis: Variable Domain Data and Partially Observed Data. See <a href="https://arxiv.org/abs/2401.05839">Amaro et al.&nbsp;(2024)</a> for the details. There are two vignettes: <a href="https://cran.r-project.org/web/packages/VDPO/vignettes/VDPO-01-introduction.html">Introduction</a> and <a href="https://cran.r-project.org/web/packages/VDPO/vignettes/VDPO-02-vd-models.html">Model fitting for variable domain data</a>.</p>
<p><a href="VDPO.png" class="lightbox" data-gallery="quarto-lightbox-gallery-17"><img src="VDPO.png" class="img-fluid" alt="Heatmap of model Beta"></a></p>
</section>
<section id="time-series" class="level3">
<h3 class="anchored" data-anchor-id="time-series">Time Series</h3>
<p><a href="https://cran.r-project.org/package=FTSgof">FTSgof</a> v1.0.0: Offers tools for the analysis of functional time series data, focusing on white noise hypothesis testing and goodness-of-fit evaluations, alongside functions for simulating data and advanced visualization techniques. See <a href="https://www.sciencedirect.com/science/article/pii/S0047259X17303512?via%3Dihub">Kokoszka et al.&nbsp;(2017)</a>, <a href="https://projecteuclid.org/journals/electronic-journal-of-statistics/volume-17/issue-1/Functional-spherical-autocorrelation--A-robust-estimate-of-the-autocorrelation/10.1214/23-EJS2112.full">Yeh et al.&nbsp;(2023)</a>, <a href="https://projecteuclid.org/journals/statistics-surveys/volume-17/issue-none/White-noise-testing-for-functional-time-series/10.1214/23-SS143.full">Kim et al.&nbsp;(2023)</a>, and <a href="https://onlinelibrary.wiley.com/doi/10.1111/jtsa.12532">Rice et al.&nbsp;(2020)</a> for the theory and the <a href="https://cran.r-project.org/web/packages/FTSgof/vignettes/FTSgof.html">vignette</a> for examples.</p>
<p><a href="FTSgof.png" class="lightbox" data-gallery="quarto-lightbox-gallery-18"><img src="FTSgof.png" class="img-fluid" alt="#d Time Series Plot"></a></p>
<p><a href="https://cran.r-project.org/package=utsf">utsf</a> v1.0.0: Implements a uniform interface for univariate time series forecasting using different regression models in an autoregressive way and provides features such as preprocessing and estimation of forecast accuracy. See the <a href="https://cran.r-project.org/web/packages/utsf/vignettes/utsf.html">vignette</a>.</p>
<p><a href="utsf.png" class="lightbox" data-gallery="quarto-lightbox-gallery-19"><img src="utsf.png" class="img-fluid" alt="Time Series with KNN Forecast"></a></p>
</section>
<section id="utilities" class="level3">
<h3 class="anchored" data-anchor-id="utilities">Utilities</h3>
<p><a href="https://cran.r-project.org/package=ctypesio">ctypesio</a> v0.1.1: Provides functions for reading and writing binary data (with files, connections, and raw vectors) using <code>C</code> type descriptions that convert data between <code>C</code> types and R types while checking for values outside the type limits, <code>NA</code> values, etc. See the vignettes <a href="https://cran.r-project.org/web/packages/ctypesio/vignettes/parse-jpeg.html">Parsing JPEG Markers</a> and <a href="https://cran.r-project.org/web/packages/ctypesio/vignettes/wave-format.html">Handcrafting a WAV file</a>.</p>
<p><a href="https://cran.r-project.org/package=GitStats">GitStats</a> v2.1.2: Provides functions to obtain standardized data from multiple <em>Git</em> services, including GitHub and GitLab. There are vignettes on <a href="https://cran.r-project.org/web/packages/GitStats/vignettes/get_repos_with_code.html">Getting Repositories</a> and <a href="https://cran.r-project.org/web/packages/GitStats/vignettes/set_hosts.html">Setting Hosts</a>.</p>
<p><a href="https://cran.r-project.org/package=mall">mall</a> v0.1.0: Enables users to run multiple Large Language Model predictions against a table. The predictions run row-wise over a specified column. It works using a one-shot prompt, along with the current row’s content. Look <a href="https://mlverse.github.io/mall/">here</a> got examples.</p>
<p><a href="https://cran.r-project.org/package=MDPIexploreR">MDPIexploreR</a> v0.2.0: Provides tools to scrape and analyze data from the <a href="https://www.mdpi.com/about/journals">MDPI journals</a>, including functions to extract metrics such as submission-to-acceptance times, article types, and plot data, and explore patterns of self-citations. See the <a href="https://cran.r-project.org/web/packages/MDPIexploreR/vignettes/introduction_MDPIexploreR.html">vignette</a>.</p>
<p><a href="MDPI.png" class="lightbox" data-gallery="quarto-lightbox-gallery-20"><img src="MDPI.png" class="img-fluid" alt="List of first six MDPI Journals"></a></p>
</section>
<section id="visualization" class="level3">
<h3 class="anchored" data-anchor-id="visualization">Visualization</h3>
<p><a href="https://cran.r-project.org/package=affiner">affiner</a> v0.1.1: Provides functions to dilate, permute, project, reflect, rotate, shear, and translate 2D and 3D points. Supports parallel projections including oblique projections such as the cabinet projection as well as axonometric projections such as the isometric projection. There is an <a href="https://cran.r-project.org/web/packages/affiner/vignettes/affiner.html">Introduction</a> and a <a href="">vignette</a> on geometry.</p>
<p><a href="affiner.png" class="lightbox" data-gallery="quarto-lightbox-gallery-21"><img src="affiner.png" style="width:50.0%"></a></p>
<p><a href="https://cran.r-project.org/package=diceplot">doceplot</a> v0.1.3: Provides functions to visualize data with more than two categorical variables and additional continuous variables that are particularly useful for exploring complex categorical data in the context of pathway analysis across multiple conditions. Look <a href="https://dice-and-domino-plot.readthedocs.io/en/latest/">here</a> for documentation.</p>
<p><a href="diceplot.png" class="lightbox" data-gallery="quarto-lightbox-gallery-22"><img src="diceplot.png" class="img-fluid" alt="Plot with 6 categorical variables"></a></p>
<p><a href="https://cran.r-project.org/package=ggcompare">ggcompare</a> v0.0.2: Add mean comparison annotations to a <code>ggplot</code> to indicate if two or more groups are significantly different. For comparisons between two groups, the p-value is calculated by t-test (parametric) or Wilcoxon rank sum test (nonparametric). For comparisons among more than two groups, the p-value is calculated by One-way ANOVA (parametric) or Kruskal-Wallis test (nonparametric). Look <a href="https://github.com/HMU-WH/ggcompare/">here</a> for examples.</p>
<p><a href="ggcompare.png" class="lightbox" data-gallery="quarto-lightbox-gallery-23"><img src="ggcompare.png" class="img-fluid" alt="Boxplots annotated with pvalues"></a></p>
<p><a href="https://cran.r-project.org/package=ggpca">ggpca</a> v0.1.2: Provides tools for creating publication-ready dimensionality reduction plots, including Principal Component Analysis (PCA), t-Distributed Stochastic Neighbor Embedding (t-SNE), and Uniform Manifold Approximation and Projection (UMAP) using the <code>ggplot2</code> framework. See the <a href="https://cran.r-project.org/web/packages/ggpca/vignettes/ggpca-examples.htm">vignette</a>.</p>
<p><a href="ggpca.png" class="lightbox" data-gallery="quarto-lightbox-gallery-24"><img src="ggpca.png" class="img-fluid" alt="PCA plot"></a></p>
<p><a href="https://cran.r-project.org/package=plotscaper">plotscaper</a> v0.2.3: Implements a framework for creating interactive figures for data exploration. All plots are automatically linked and support several kinds of interactive features, including selection, zooming, panning, and parameter manipulation. There is an <a href="https://cran.r-project.org/web/packages/plotscaper/vignettes/plotscaper.html">Introduction</a> and four additional vignettes, including <a href="https://cran.r-project.org/web/packages/plotscaper/vignettes/available-interactions.html">Available Interactions</a> and <a href="https://cran.r-project.org/web/packages/plotscaper/vignettes/layout.html">Layout</a>.</p>
<p><a href="plotscaper.png" class="lightbox" data-gallery="quarto-lightbox-gallery-25"><img src="plotscaper.png" class="img-fluid" alt="Four interlinked plots"></a></p>
<p><a href="https://cran.r-project.org/package=vchartr">vchartr</a> v0.1.3: Provides an <code>htmlwidgets</code> interface to <a href="https://www.visactor.io/vchart">VChart</a>, a cross-platform charting library and expressive data storyteller. See the <a href="https://cran.r-project.org/web/packages/vchartr/vignettes/vchartr.html">vignette</a>.</p>
<p><a href="vchart.png" class="lightbox" data-gallery="quarto-lightbox-gallery-26"><img src="vchart.png" class="img-fluid" alt="Examples of charts"></a></p>
</section>
</div>
</div>
<hr>
<section id="comments" class="level2 unlisted">
<h2 class="unlisted anchored" data-anchor-id="comments">Comments</h2>
<bluesky-comments-section post="https://bsky.app/profile/rworks.bsky.social/post/3lbwuyobgx22e"></bluesky-comments-section>


</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp("https:\/\/rworks\.dev");
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->
<footer class="footer"><div class="nav-footer"><div class="nav-footer-center"><div class="toc-actions"><ul><li><a href="https://github.com/r-community-works/rworks-website/blob/main/posts/october-2024-top-40-new-cran-packages/index.qmd" class="toc-action"><i class="bi bi-github"></i>View source</a></li><li><a href="https://github.com/r-community-works/rworks-website/issues/new" class="toc-action"><i class="bi empty"></i>Report an issue</a></li></ul></div></div></div></footer><script>var lightboxQuarto = GLightbox({"closeEffect":"zoom","descPosition":"bottom","loop":false,"openEffect":"zoom","selector":".lightbox"});
(function() {
  let previousOnload = window.onload;
  window.onload = () => {
    if (previousOnload) {
      previousOnload();
    }
    lightboxQuarto.on('slide_before_load', (data) => {
      const { slideIndex, slideNode, slideConfig, player, trigger } = data;
      const href = trigger.getAttribute('href');
      if (href !== null) {
        const imgEl = window.document.querySelector(`a[href="${href}"] img`);
        if (imgEl !== null) {
          const srcAttr = imgEl.getAttribute("src");
          if (srcAttr && srcAttr.startsWith("data:")) {
            slideConfig.href = srcAttr;
          }
        }
      } 
    });
  
    lightboxQuarto.on('slide_after_load', (data) => {
      const { slideIndex, slideNode, slideConfig, player, trigger } = data;
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(slideNode);
      }
    });
  
  };
  
})();
          </script>




</body></html>